sql/information_schema/columns_table.go

```diff
@@ -0,0 +1,365 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package information_schema
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+)
+
+// ColumnsNode wraps the information_schema.columns table as a way to resolve column defaults.
+type ColumnsNode struct {
+	plan.UnaryNode
+
+	// columnNameToColumn maps the name of a column (databaseName.tableName.columnName) with a non nil default value
+	// to its column object.
+	columnNameToColumn map[string]*sql.Column
+
+	// defaultToColumn maps the pointer of a sql.ColumnDefault value back to its original Column Object.
+	defaultToColumn map[*sql.ColumnDefaultValue]*sql.Column
+}
+
+var _ sql.Node = (*ColumnsNode)(nil)
+var _ sql.Expressioner = (*ColumnsNode)(nil)
+var _ sql.Nameable = (*ColumnsNode)(nil)
+
+// CreateNewColumnsNode returns a new ColumnsNode.
+func CreateNewColumnsNode(child sql.Node, tableToColumnsWithDefaultValue map[string]*sql.Column) *ColumnsNode {
+	return &ColumnsNode{
+		UnaryNode:          plan.UnaryNode{Child: child},
+		columnNameToColumn: tableToColumnsWithDefaultValue,
+	}
+}
+
+// Resolved implements the sql.Node interface.
+func (c *ColumnsNode) Resolved() bool {
+	return c.Child.Resolved() && c.defaultToColumn != nil
+}
+
+// String implements the sql.Node interface.
+func (c *ColumnsNode) String() string {
+	return fmt.Sprintf("ColumnsNode(%s)", c.Child.String())
+}
+
+// Schema implements the sql.Node interface.
+func (c *ColumnsNode) Schema() sql.Schema {
+	return c.Child.Schema()
+}
+
+// RowIter implements the sql.Node interface.
+func (c *ColumnsNode) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	ct := getColumnsTable(c)
+
+	colToDefaults := make(map[string]*sql.ColumnDefaultValue)
+	for colName, col := range c.columnNameToColumn {
+		colToDefaults[colName] = col.Default
+	}
+
+	ct.WithColumnNametoDefaultMap(colToDefaults)
+
+	return c.Child.RowIter(ctx, row)
+}
+
+// WithChildren implements the sql.Node interface.
+func (c *ColumnsNode) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(c, len(children), 1)
+	}
+
+	nc := *c
+	nc.Child = children[0]
+	return &nc, nil
+}
+
+// CheckPrivileges implements the sql.Node interface.
+func (c *ColumnsNode) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	return c.Child.CheckPrivileges(ctx, opChecker)
+}
+
+// Expressions implements the sql.Expressioner interface.
+func (c *ColumnsNode) Expressions() []sql.Expression {
+	c.defaultToColumn = make(map[*sql.ColumnDefaultValue]*sql.Column)
```

Comment:
Don't use maps for any of this, it makes everything way more complicated

Just consider all the defaults for all the tables in sorted name order, and use the same order here and in WithExpressions and you'll be fine

Reply:
I considered this but worried about inefficiencies. Expressions() gets called a lot so Iterating through all columns in database seems expensive.

Reply:
This map is also essential in the resolve_column_defaults step where we need to actual column associated with the default to resolve. Moreover, we need the other columnNameToColumn map as a way to pass down the map down to the ultimate row iter

Reply:
nvm figured it out smh